backwardcompatibilityml/helpers/training.py

```diff
@@ -504,36 +520,34 @@ def evaluate_model_performance_and_compatibility_on_dataset(h1, h2, dataset, per
         the trust compatibility score of h2 with respect to h1, and
         the error compatibility score of h2 with respect to h1.
     """
-    number_of_batches = len(dataset)
     h1_dataset_error_instance_ids = []
     h2_dataset_error_instance_ids = []
     h1_and_h2_dataset_error_instance_ids = []
-    h2_dataset_error_instance_ids_by_class = {}
+    h1h2_dataset_incompatible_instance_ids_by_class = {}
     classes = set()
     all_error_instances = []
     for batch_ids, data, target in dataset:
         classes = classes.union(target.tolist())
         h1_error_count_batch, h2_error_count_batch, h1_and_h2_error_count_batch =\
             get_model_error_overlap(h1, h2, batch_ids, data, target, device=device)
-        h2_error_instance_ids_by_class =\
-            get_error_instance_ids_by_class(h2, batch_ids, data, target, device=device)
         all_errors = get_all_error_instance_indices(
-            h1, h2, data, target, device=device)
+            h1, h2, batch_ids, data, target, device=device)
+        h1h2_incompatible_ids_by_class = get_incompatible_instances(all_errors, batch_ids, target)
         all_error_instances += all_errors
         h1_dataset_error_instance_ids += h1_error_count_batch
         h2_dataset_error_instance_ids += h2_error_count_batch
         h1_and_h2_dataset_error_instance_ids += h1_and_h2_error_count_batch
-        for class_label, error_instance_ids in h2_error_instance_ids_by_class.items():
-            if class_label in h2_dataset_error_instance_ids_by_class:
-                h2_dataset_error_instance_ids_by_class[class_label] += error_instance_ids
+        for class_label, incompatible_instance_ids in h1h2_incompatible_ids_by_class.items():
```

Comment:
I'd suggest folding all these calculations into the helper function you added, and just having that function return a dictionary mapping the classes to incompatible points.